Sources/SwiftPlot/HistogramSeries.swift

```diff
@@ -11,6 +13,7 @@ public struct HistogramSeries<T> where T: Comparable {
                 color: Color,
                 histogramType: HistogramSeriesOptions.HistogramType) {
         self.data = data.sorted()
+        self.isSorted = true
```

Comment:
But while setting the data we sort it. 

Self.data = data.sorted()

Reply:
We happen to do that today, but it's fragile to be relying on histogram always pre-sorting the data - maybe somebody will add a method to histogram and forget to sort the data first. There would be no error, just worse performance.

Also, this initialiser is public and `Histogram.histogramSeries` is publicly settable; there is nothing stopping anybody replacing the instance entirely with `myHistogram.histogramSeries = HistogramSeries(data: [4, 3, 2, 1], ...)`.

Reply:
> Also, this initialiser is public and Histogram.histogramSeries is publicly settable; there is nothing stopping anybody replacing the instance entirely with myHistogram.histogramSeries = HistogramSeries(data: [4, 3, 2, 1], ...). If we assume that it's always sorted, we would display an invalid graph.

Ok I think you might be misunderstanding where `self.isSorted = true` is defined. The example you gave would lead to correct behaviour, and we would not display an invalid graph. This is because we sort inside the `HistogramSeries` initialiser at the same time that we set `isSorted = true`, not in `Histogram`'s initialiser or anywhere else outside `HistogramSeries`'s initialiser. So `HistogramSeries(data: [4, 3, 2, 1], ...)` sorts the data before storing it and then sets `isSorted = true`. Anyone that touches `data` directly afterwards would invalidate this flag, thus setting `isSorted = false` implicitly.
I do agree on having a separate setter with a parameter to ask if the data is sorted though.

I guess the question is if we want to keep the current behaviour of master, which is to sort any data that comes through the initialiser of `HistogramSeries` or if we want to stop doing that every time and instead expose the `isSorted` parameter through that same initialiser, giving the option of pre-sorting to `Histogram` and others while giving the ability, at instance creation, to inform `HistogramSeries` that the data passed in is already sorted.

Hopefully this makes sense and I did not misunderstand anyone!

Reply:
> This is because we sort inside the HistogramSeries initialiser

Ah, my mistake - it's actually on the very line above this one 🤦‍♂

I guess in the long run it makes to give the user an option to not sort the data (since apparently that's something we support), but for now it's fine not to do the initialiser parameter.

I'm still not entirely happy with assuming that any modification disrupts the data's sorting. This isn't an entirely easy thing to solve though. If we replaced direct setting with a method, things like `.append` would become unavailable. The only way to keep those methods is to use a closure:

```swift
mutating func modifyData(preservesSorting: Bool, _ closure: (inout [T])->Void) {
  closure(&data)
  isSorted = !preservesSorting
}
```

But that's not entirely obvious, either - because we automatically sorted the data for the user, they might not be aware that by "preserves" sorting, we mean that they won't un-sort previously sorted data. Which is to say: if they gave us unsorted data, then make some random edits, they might say that they "preserve" sorting because the data wasn't sorted before and still isn't sorted afterwards.

Reply:
Actually - there is probably an even better option: don't sort the data.

There are two main reasons for this:

- **Predictability**: As a user, I just created a histogram with some data. Now I want to access or modify some of that data, but it has unexpectedly changed from what I gave it. Everything is in a different location, so I can't reliably process it.

- **Performance**: Sorting is an `O(n log(n))` operation, which is more than we save by having the data sorted. Also, it can lead to additional copies of the data (another `O(n)` operation). And then if a user modifies the data after creation, we throw away all that work we did by sorting anyway. We made it harder for them to use, and it wasn't even for a good reason.

Reply:
I can see how it can quickly become hairy... My initial implementation was based on the fact that the data was already being sorted every time in the initialiser, so I though "why not take advantage of that fact when possible?".My aim was not to introduce a new `isSorted` property that everyone else now would have to be careful with so I decided to not include any public api that exposed this, and only added the necessary code to make sure that the algorithm doesn't break (the `didSet`). In a way `isSorted` doesn't really get through the idea I wanted, it should be more like `is(Definitely|Undeniably)Sorted` since I only cared that the fast algorithm never runs when `data` actually isn't sorted, with no big harm if `data` is sorted but the algorithm that runs is the slower one.

I don't know if you want to continue discussing solutions to possible public facing apis here, that don't run into the disadvantages you mention above.

What do you think of doing the following for right now:
- set `isSorted = false` on any modification to `data`
- no `private(set)` so we can still use `.append` and such directly
- we could still have a separate public setter (`setData(_:sorted:)`) to give performance minded users the ability to say if their new data is sorted or not (this could also come later)
- keep sorting the data and setting `isSorted = true` in the initialiser without exposing an `isSorted` parameter on it

Then in another PR/Issue we could discuss how we could expose the standard apis of `data` while letting the user specify sortedness without the drawbacks you mentioned. This includes discussing the third point above (`setData`) if we decide to not include it right now in this PR.

Reply:
I think first we should consider if sorting the data is even worthwhile. It may well be hurting performance rather than helping it.

We can then consider if it makes sense, if the user already happens to have sorted data, to allow them to mention that and benefit from a faster algorithm.

Reply:
oh sorry, I took my time writing and only now I see another comment of yours came through before mine

Reply:
> Actually - there is probably an even better option: don't sort the data.
> 
> There are two main reasons for this:
> 
> * **Usability**: As a user, I just created a histogram with some data. Now I want to access or modify some of that data, but it has unexpectedly changed from what I gave it. Everything is in a different location, so I can't reliably process it.
> * **Performance**: Sorting is an `O(n log(n))` operation, which is more than we save by having the data sorted. Also, it can lead to additional copies of the data (another `O(n)` operation). And then if a user modifies the data after creation, we throw away all that work we did by sorting anyway. We made it harder for them to use, and it wasn't even for a good reason.
> 
> Would you mind benchmarking it and seeing what happens if we just never sort the data?

ok yes that makes sense, I will benchmark it

Reply:
> I think first we should consider if sorting the data is even worthwhile. If our algorithm didn't even need it in the first place, that's a clear an obvious removal of unnecessary work, with no API awkwardness to worry about.
> 
> We can then consider if it makes sense, if the user already _happens_ to have sorted data, to allow them to mention that and benefit from a faster algorithm.

That also makes sense 👍

Reply:
Option 1 would differ from the other plots, and ultimately it all depends on whether there _are_ any such optimisations (histograms are a well-explored domain, and I couldn't find anything), what the cost is to convert the data to that format, and what the gains are. Given that any copy will cost us _at least_ `O(n)`, I think it's unlikely that any data structure exists that would give us back the cost.

> not because it would be the only option that would allow us sorting the data whenever we want

Like I said, it's not clear that sorting the data actually _is_ a win, whenever we do it. It costs up to `O(n log n)`, all so we can turn another `O(n log n)` operation to only `O(n)`. We should see an overall performance boost by cutting all of that out and only doing a single `O(n log n)` operation on the unsorted data. In other words: it should be _even faster_ than what you have right now.

If the user already has sorted data from somewhere - fine, we can use that information. That's (essentially) zero-cost for us, and you've shown that you can extract gains from that knowledge.

If we encounter some compelling reason to use another data structure, we'll consider what to do at that time. As it is, I don't see the harm.

Reply:
Yes, I completely agree with you, what I meant with that was that my opinion to choose the first option wasn't based on being able to sort or not ahah. And with that I have the results of the experiment!

First to make sure I am not making some mistake please validate what I did:
Create new temp initialiser that does not sort and sets `isSorted = false`:

```swift
public init(dataNotSorted: [T],
            bins: Int,
            label: String,
            color: Color,
            histogramType: HistogramSeriesOptions.HistogramType) {
    self.data = dataNotSorted
    self.isSorted = false
    self.bins = bins
    self.label = label
    self.color = color
    histogramSeriesOptions.histogramType = histogramType
}
```

Then the two tests:

```swift
// time of fast algorithm + necessary sort
func testPerformanceHistogramCreateSeriesAndRecalculateBins_Sorting() throws {
    let histogram = Histogram<Float>(isNormalized: false, enableGrid: false)
    
    measure {
        let histogramSeries = HistogramSeries(data: histogram_step_values, bins: 50, label: "HISTOGRAM PERFORMANCE `addSeries` & `recalculateBins`", color: .black, histogramType: .bar)
        histogram.testRecalculateBins(series: histogramSeries, binStart: 40, binEnd: 160, binInterval: (160-40)/Float(histogram.histogramSeries.bins))
    }
}
// time of slow algorithm
func testPerformanceHistogramCreateSeriesAndRecalculateBins_NonSorting() throws {
    let histogram = Histogram<Float>(isNormalized: false, enableGrid: false)
    
    measure {
        let histogramSeries = HistogramSeries(dataNotSorted: histogram_step_values, bins: 50, label: "HISTOGRAM PERFORMANCE `addSeries` & `recalculateBins`", color: .black, histogramType: .bar)
        histogram.testRecalculateBins(series: histogramSeries, binStart: 40, binEnd: 160, binInterval: (160-40)/Float(histogram.histogramSeries.bins))
    }
}
```

assuming I'm not overlooking anything the results are:
<img width="830" alt="Screenshot 2020-01-29 at 20 10 16" src="https://user-images.githubusercontent.com/278662/73393340-60ee8b00-42d3-11ea-8903-58ad11f2c8ca.png">

which is a pretty clear win for not sorting in the first place! So it is now time to simplify 😂.

Nonetheless I still think option 1 would make sense, not exposing the underlying array at all (which in my opinion is an implementation detail), unless there is a compelling enough reason, and reconsider the same for all the other plot types, but I haven't actually looked too hard at the rest nor have I though enough about what that would mean overall.

Reply:
I don't think the difference between the fast and slow algorithms, when a user gives an already sorted array, justifies the added complexity of two algorithms and maintaining an need comes we can think about giving the user the option of giving sorted data.

Reply:
The above discussion has been illuminating for me! I never thought about the cost of sorting the data prior to binning.

What does **(1% worse)** actually mean? What is that time being compared to?

One more thing unrelated to this dicussion: matplotlib return the bins and the histogram values. Should we consider adding a function to do something like this?

Reply:
That time is being compared to a baseline set previously, in my case by measuring. The 1% worse just means that this time it was 1% slower than the baseline.

Regarding exposing the bins and the histogram values, in my opinion we should come up with use cases for that and weight if it makes sense/is desirable first.
<img width="249" alt="Screenshot 2020-01-29 at 20 50 24" src="https://user-images.githubusercontent.com/278662/73396298-fe988900-42d8-11ea-8177-d4f5efe79da1.png">

Reply:
We can consider hiding the underlying array if/when a need arises. Maybe it will be for this plot, maybe for some other plot. I'm open to it if there's a compelling use-case.

In the mean time, those tests look good.

Reply:
Regarding exposing the bins and histogram weights, I talked to someone who works in ML. These values come in useful during classification tasks where you need to categorize continuous values into discrete values.

Reply:
@KarthikRIyer that's more-or-less the same data that's in Histogram's `DrawingData` type, isn't it? We could definitely support querying that directly.

We could also expose that through something similar to MATLAB's [`histcounts` API](https://de.mathworks.com/help/matlab/ref/histcounts.html).

This gets more interesting if we introduce algorithms to automatically determine a good number of bins. MATLAB provides [lots of options](https://de.mathworks.com/help/matlab/ref/histcounts.html#buijikm-1-BinMethod), and there's lots of research with alternative algorithms.

Reply:
@karwa yes.

Yeah, this idea for auto number of bins is very interesting! We can definitely get that implemented. I'd love to work on this, but I'm not sure how soon I can get to it.